### PR TITLE
Fix CodeLens positioned inside preceding code blocks

### LINKED ISSUE #15618: VS Code: Code Lens shown in weird place
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Describe the bug

The code lens for running resources is sometimes showing up in weird spots.  Take the following for example

Image

### Expected Behavior

I'd expect the code lens to be immediately on the line above the `builder.AddContainer()` call, not within the if statement above it.

### Steps To Reproduce

Add the following code, to an app host in VS Code, then run the app host with the extension
```cs
if (false)
{
	throw new Exception("This should not work");
}

builder.AddContainer("nginx", "nginx");
```

### Exceptions (if any)

_No response_

### .NET Version info

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by adamint**
## Description

Fix CodeLens being positioned inside preceding code blocks instead of on the resource declaration line.

**Root cause:** `_findStatementStartLine()` in both the C# and JS/TS AppHost parsers walked backwards from a resource match looking for `;` or `{` as statement delimiters, but ignored `}`. When a code block like `if (...) { ... }` preceded a resource call, the algorithm walked past the `}` into the block and stopped at its `{`, placing the CodeLens inside the preceding block.

**Fix:** Add `}` as a statement delimiter in the backward walk so that closing braces correctly bound the start of the current statement.

**Also fixes:** `tsconfig.json` updated to include `mocha` types (resolves "Cannot find name 'test'" errors in VS Code) and `skipLibCheck: true` (resolves duplicate MCP type definition conflicts between `@types/vscode` and the proposed API file).

**Validation:** 26 new tests (13 C#, 13 JS/TS) covering: preceding if blocks, nested braces, blocks with semicolons, comments between blocks and resources (line comments, block comments, mixed comment styles), multi-line fluent chains, try/catch blocks, single-line blocks, empty blocks, for loops, multiple resources interspersed with blocks, and comments before fluent chains after blocks. All existing tests continue to pass.

Fixes #15618

## Checklist

- Is this feature complete?
  - [x] Yes. Ready to ship.
  - [ ] No. Follow-up changes expected.
- Are you including unit tests for the changes and scenario tests if relevant?
  - [x] Yes
  - [ ] No
- Did you add public API?
  - [ ] Yes
    - If yes, did you have an API Review for it?
      - [ ] Yes
      - [ ] No
    - Did you add `<remarks />` and `<code />` elements on your triple slash comments?
      - [ ] Yes
      - [ ] No
  - [x] No
- Does the change make any security assumptions or guarantees?
  - [ ] Yes
    - If yes, have you done a threat model and had a security review?
      - [ ] Yes
      - [ ] No
  - [x] No
- Does the change require an update in our Aspire docs?
  - [ ] Yes
    - Link to `aspire.dev` issue:
      - [New issue](https://github.com/microsoft/aspire.dev/issues/new)
  - [x] No

### CODE REVIEW

File: extension/src/editor/parsers/csharpAppHostParser.ts [hunk 1]
```diff
@@ -58,7 +58,7 @@ class CSharpAppHostParser implements AppHostResourceParser {
         let i = matchIndex - 1;
         while (i >= 0) {
             const ch = text[i];
-            if (ch === ';' || ch === '{') {
+            if (ch === ';' || ch === '{' || ch === '}') {
                 break;
```
[the remark this answers is not in the document]
Comment (adamint) [reply extension/src/editor/parsers/csharpAppHostParser.ts]: Good catch — updated the doc comments and call-site comments to mention `}` in both parsers.

File: extension/src/editor/parsers/jsTsAppHostParser.ts [hunk 2]
```diff
@@ -59,7 +59,7 @@ class JsTsAppHostParser implements AppHostResourceParser {
         let i = matchIndex - 1;
         while (i >= 0) {
             const ch = text[i];
-            if (ch === ';' || ch === '{') {
+            if (ch === ';' || ch === '{' || ch === '}') {
                 break;
```
[the remark this answers is not in the document]
Comment (adamint) [reply extension/src/editor/parsers/jsTsAppHostParser.ts]: Fixed — doc comments and call-site comments updated to mention `}` in the JS/TS parser as well.

File: extension/src/editor/parsers/csharpAppHostParser.ts [hunk 3]
```diff
@@ -58,7 +58,7 @@ class CSharpAppHostParser implements AppHostResourceParser {
...
             }
```
[the remark this answers is not in the document]
Comment (adamint) [reply extension/src/editor/parsers/csharpAppHostParser.ts]: Great edge case — fixed. The comment-skip loop now also skips lines starting with `}` (after trimming), since a `}` line is never a valid statement start. Added a test for `} // end if` directly before a resource call. The CodeLens now correctly skips past it.

File: extension/src/editor/parsers/jsTsAppHostParser.ts [hunk 4]
```diff
@@ -59,7 +59,7 @@ class JsTsAppHostParser implements AppHostResourceParser {
...
             }
```
[the remark this answers is not in the document]
Comment (adamint) [reply extension/src/editor/parsers/jsTsAppHostParser.ts]: Fixed — same as the C# parser, the comment-skip loop now also skips lines starting with `}`. Added a corresponding test for `} // end if` in the JS/TS suite.

File: extension/tsconfig.json [hunk 5]
```diff
@@ -8,7 +8,12 @@
 		"sourceMap": true,
 		"rootDir": "src",
 		"strict": true,   /* enable all strict type-checking options */
-		"outDir": "out"
+		"outDir": "out",
+		"skipLibCheck": true,
+		"types": [
+			"mocha",
+			"node"
```
[the remark this answers is not in the document]
Comment (adamint) [reply extension/tsconfig.json]: Not needed — `@types/vscode` is resolved via explicit `import * as vscode from 'vscode'` statements in each source file, not through ambient `@types` inclusion. The `types` array only controls which `@types/*` packages provide **ambient global** declarations (like `mocha`'s `test`/`suite` globals). I verified compilation succeeds without `"vscode"` in the `types` array.

File: extension/src/editor/parsers/csharpAppHostParser.ts [hunk 6]
```diff
@@ -51,14 +51,14 @@ class CSharpAppHostParser implements AppHostResourceParser {
...
-     * Stops at the previous ';', '{', or start of file, then returns the first non-comment,
+     * Stops at the previous ';', '{', '}', or start of file, then returns the first non-comment,
      * non-blank line after that delimiter.
      */
     private _findStatementStartLine(text: string, matchIndex: number, document: vscode.TextDocument): number {
         let i = matchIndex - 1;
         while (i >= 0) {
             const ch = text[i];
-            if (ch === ';' || ch === '{') {
+            if (ch === ';' || ch === '{' || ch === '}') {
```
Comment (sebastienros) [inline extension/src/editor/parsers/csharpAppHostParser.ts]: This still breaks existing fluent chains where the `Add*` call comes **after** a callback block in the same statement. For example, `playground/TestShop/TestShop.AppHost/AppHost.cs` has `.WithPgAdmin(resource => { ... }).AddDatabase("catalogdb")`, and `playground/waitfor/WaitForSandbox.AppHost/AppHost.cs` has `.RunAsContainer(c => { ... }).AddDatabase("db")`. In both cases the new `}` stop condition makes `_findStatementStartLine` stop at the callback closing brace, so `statementStartLine` becomes the `.AddDatabase(...)` line instead of the top of the fluent chain.

That means the CodeLens is still misplaced for real AppHost code after merge, and it regresses the `statementStartLine` contract from "first line of the full statement chain" to "line after the last callback block". The fix likely needs to distinguish a `}` that closes a *previous* statement from one that belongs to the current fluent call (for example by tracking nesting / scanning for the previous real statement boundary), and this needs a focused regression test for these callback-chain shapes. The same issue is duplicated in `jsTsAppHostParser.ts`.

File: extension/src/test/parsers.test.ts
Comment (JamesNK) [inline extension/src/test/parsers.test.ts:1]: None of the new tests cover the scenario where `}` appears inside the *same* fluent statement — e.g., a callback lambda in a chain:

```csharp
var catalogDb = builder.AddPostgres("postgres")
                       .WithPgAdmin(resource => { ... })
                       .AddDatabase("catalogdb");
```

All 26 new tests only place `}` in *separate preceding statements* (if-blocks, try-catch, for-loops), so they don't catch the regression @sebastienros flagged where the backward walk stops at the callback's `}` instead of reaching `var catalogDb`. A test using this shape (for both C# and JS/TS) would have caught it immediately and should be added alongside whatever fix is chosen for the nesting issue.

File: extension/tsconfig.json [hunk 7]
```diff
@@ -8,7 +8,12 @@
 		"sourceMap": true,
 		"rootDir": "src",
 		"strict": true,   /* enable all strict type-checking options */
-		"outDir": "out"
+		"outDir": "out",
+		"skipLibCheck": true,
```
Comment (JamesNK) [inline extension/tsconfig.json]: `skipLibCheck: true` disables type-checking for *all* `.d.ts` files, not just the conflicting MCP type definitions. This means type errors in any dependency's declarations (or the proposed VS Code API file) will be silently ignored at compile time.

A more targeted approach would preserve type safety for production code — for example, a separate `tsconfig.test.json` that extends the base config and adds `mocha` types without needing `skipLibCheck`, or resolving the duplicate definitions directly.

File: extension/src/editor/parsers/csharpAppHostParser.ts [hunk 8]
```diff
@@ -71,10 +71,10 @@ class CSharpAppHostParser implements AppHostResourceParser {
...
         let line = document.positionAt(start).line;
         const matchLine = document.positionAt(matchIndex).line;
-        // Skip lines that are comments (// or /* or * continuation)
+        // Skip lines that are closing braces or comments (// or /* or * continuation)
         while (line < matchLine) {
             const lineText = document.lineAt(line).text.trimStart();
-            if (lineText.startsWith('//') || lineText.startsWith('/*') || lineText.startsWith('*')) {
+            if (lineText.startsWith('}') || lineText.startsWith('//') || lineText.startsWith('/*') || lineText.startsWith('*')) {
                 line++;
             } else {
```
Comment (JamesNK) [inline extension/src/editor/parsers/csharpAppHostParser.ts]: The `lineText.startsWith('}')` check will also skip lines like `} else {` or `} catch (e) {`, where the `{` on the same line opens a new block. If such a line existed between a block and a resource call, the opener `{` wouldn't act as a delimiter — the entire line would just be skipped. This is a minor edge case (unlikely before resource declarations in practice), but the same pattern is duplicated in `jsTsAppHostParser.ts`.

**Review by JamesNK — COMMENTED** [review]
3 issues found: missing test coverage for callback-in-chain regression (1), `skipLibCheck: true` suppressing all `.d.ts` type-checking (1), minor edge case in `}` line-skip logic (1).

File: extension/tsconfig.json (hunk 7 above)
Comment (adamint) [reply extension/tsconfig.json]: Agreed — reverted. Both `skipLibCheck` and the `types` array have been removed; compilation succeeds without them. The mocha ambient types and MCP type definition conflicts are pre-existing issues that can be addressed separately (e.g. a `tsconfig.test.json` for mocha globals, and resolving the proposed API duplicate definitions directly).

File: extension/src/editor/parsers/csharpAppHostParser.ts (hunk 6 above)
Comment (adamint) [reply extension/src/editor/parsers/csharpAppHostParser.ts]: Fixed. The backward walk now tracks brace nesting: when it encounters `}`, it finds the matching `{` and checks whether `=>` precedes it. If so, the block is recognized as a lambda body within the current fluent chain and is skipped over. Otherwise it's treated as a separate statement boundary.

This correctly handles patterns like `.WithPgAdmin(resource => { ... }).AddDatabase("catalogdb")` and `.RunAsContainer(c => { ... }).AddDatabase("db")` — the CodeLens now reaches the top of the fluent chain through callback lambdas.

Added tests for both shapes (C# and JS/TS).

File: extension/src/editor/parsers/csharpAppHostParser.ts (hunk 8 above)
Comment (adamint) [reply extension/src/editor/parsers/csharpAppHostParser.ts]: Fixed. The `lineText.startsWith('}')` check has been replaced with a regex `/^\}\s*(\/\/.*)?$/` that only matches lines where `}` is followed by optional whitespace and an optional `//` comment. Lines like `} else {` and `} catch (e) {` are no longer incorrectly skipped.

Added a test for `} else { ... }` preceding a resource call to verify correctness in both parsers.

File: extension/src/test/parsers.test.ts
Comment (adamint) [reply extension/src/test/parsers.test.ts:1]: Added. New tests cover:
- `statementStartLine reaches top of fluent chain through callback lambda` — `.WithPgAdmin(resource => { ... }).AddDatabase("catalogdb")` shape (C# and JS/TS)
- `statementStartLine reaches top of fluent chain through RunAsContainer callback` — `.RunAsContainer(c => { ... }).AddDatabase("db")` shape (C# and JS/TS)

Both verify that `statementStartLine` correctly points to the `var catalogDb` / `const catalogDb` line, not to the line after the callback's `}`.

**Review by JamesNK — APPROVED** [review]
There is some duplication. It would be good if parsing can be reused where possible. e.g. C#, TS and Java probably share some parsing logic because they're all C syntax languages.
